app/models/data_repo_client.rb

```diff
@@ -0,0 +1,199 @@
+      'x-app-id' => "single-cell-portal",
+      'x-domain-id' => "#{ENV['HOSTNAME']}"
+    }
+
+    # process request
+    begin
+      response = RestClient::Request.execute(method: http_method, url: path, payload: payload, headers: headers)
+      # handle response using helper
+      handle_response(response)
+    rescue RestClient::Exception => e
+      current_retry = retry_count + 1
+      context = " encountered when requesting '#{path}', attempt ##{current_retry}"
+      log_message = "#{e.message}: #{e.http_body}; #{context}"
+      Rails.logger.error log_message
+      retry_time = retry_count * ApiHelpers::RETRY_INTERVAL
+      sleep(retry_time)
+      # only retry if status code indicates a possible temporary error, and we are under the retry limit and
+      # not calling a method that is blocked from retries
+      if should_retry?(e.http_code) && retry_count < ApiHelpers::MAX_RETRY_COUNT
+        process_api_request(http_method, path, payload: payload, retry_count: current_retry)
+      else
+        # we have reached our retry limit or the response code indicates we should not retry
+        ErrorTracker.report_exception(e, self.issuer, {
+          method: http_method, url: path, payload: payload, retry_count: retry_count
+        })
+        error_message = parse_response_body(e.message)
+        Rails.logger.error "Retry count exceeded when requesting '#{path}' - #{error_message}"
+        raise e
+      end
+    end
+  end
+
+  ##
+  # API ENDPOINT BINDINGS
+  ##
+
+  ##
+  # Datasets
+  ##
+
+  # fetch all available datasets
+  #
+  # * *params*
+  #   - +direction+ (String) => Sort direction (default: desc)
+  #   - +filter+ (String) => Filter the results where this string is a case insensitive match in the name or description.
+  #   - +limit+ (Integer) => The numbers datasets to retrieve
+  #   - +offset+ (Integer) => The number of datasets to skip before when retrieving the next page
+  #   - +sort+ (String) => The field to use for sorting: (name, description, created_date)
+  #
+  # * *returns*
+  #   - (Array<Hash>) => Array of dataset JSON attributes
+  def datasets(direction: 'desc', filter: nil, limit: 100, offset: 0, sort: 'name')
+    validate_argument(:direction, direction, SORT_DIRECTIONS)
+    validate_argument(:sort, sort, SORT_OPTIONS)
+    query_opts = merge_query_options({direction: direction, filter: filter, limit: limit, offset: offset, sort: sort})
+    path = api_root + '/api/repository/v1/datasets' + query_opts
+    process_api_request(:get, path)
+  end
+
+  # fetch a single dataset
+  #
+  # * *params*
+  #   - +dataset_id+ (String: UUID) => Dataset UUID
+  #   - +include+ (Array) => Dataset fields to return from DATASET_INCLUDE_FIELDS (default: SCHEMA,PROFILE,DATA_PROJECT,STORAGE)
+  #
+  # * *returns*
+  #   - (Hash) => Hash of dataset JSON attributes
+  def dataset(dataset_id, include: ALL_DATASET_FIELDS)
+    path = api_root + "/api/repository/v1/datasets/#{dataset_id}"
```

Comment:
We wouldn't be able to do that because of the DRS endpoints: https://jade.datarepo-dev.broadinstitute.org/swagger-ui.html#/DataRepositoryService

Those will be needed to get information about the location of files.

Reply:
Oof, nice catch on something of an API gotcha!